CMakeLists.txt

```diff
@@ -139,10 +139,11 @@ if (MFEM_USE_CUDA)
       set(CMAKE_CUDA_HOST_LINK_LAUNCHER ${CMAKE_CXX_COMPILER})
    endif()
    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} ${CUDA_FLAGS}")
+   find_package(CUDAToolkit REQUIRED)
    set(CUSPARSE_FOUND TRUE)
-   set(CUSPARSE_LIBRARIES "cusparse")
+   get_target_property(CUSPARSE_LIBRARIES CUDA::cusparse LOCATION)
    set(CUBLAS_FOUND TRUE)
-   set(CUBLAS_LIBRARIES "cublas")
+   get_target_property(CUBLAS_LIBRARIES CUDA::cublas LOCATION)
```

Comment:
This is not caused by this PR, however, it is a related issue: `cublas` is not needed by MFEM directly; it is needed by HYPRE when it is built with CUDA. We add the HYPRE CUDA dependencies here: https://github.com/mfem/mfem/blob/ecfab8fe9f6f01dae8fc9fc59783bf9581eb3780/config/defaults.cmake#L109-L113
It maybe best to move the `cublas` dependency there. What do you think?

Reply:
Yeah, I noticed this was a bit inconsistent. For HIP builds, [`FindHYPRE.cmake`](https://github.com/mfem/mfem/blob/master/config/cmake/modules/FindHYPRE.cmake#L20C23-L22C29) is taking charge of adding the `rocsparse` and `rocrand` dependencies, which is also not great let alone inconsistent with the CUDA builds. There are cases where HYPRE is actually not build with `cublas` nor `curand`, so I think we should:

- Remove `cublas` as a dependency of MFEM from `CMakeLists.txt`.
- Move the handling of `rocsparse` and `rocrand` to `default.cmake` from `FindHYPRE.cmake` and handle them like the CUDA case, where the user can override the values as needed at configure time by specifying `HYPRE_REQUIRED_LIBRARIES` directly based on how they built Hypre.

Does that sound like a good plan?

Reply:
I prefer if we move the HYPRE+CUDA handling to be automated similar to HYPRE+HIP. If you are concerned that auto-detection may fail (or link with unnecessary libs), we can add an option to overwrite it.

Reply:
For this PR, we can leave the HYPRE+HIP and HYPRE+CUDA handling as it is. Let's just move `cublas` from the main CMake file to `defaults.cmake`. What do you think?

Reply:
Since it is an easy change, I've gone ahead and made the Hypre handling of cuSPARSE and cuRAND dependencies match that for HIP. I still think this will add unnecessary linkage when the user has not built HYPRE with `--enable-curand` or `--enable-cusparse` (or the ROCm counterparts), but don't see a good way of automatically handling this while also allowing the user to override the defaults. The autodetection should never fail since these are standard parts of the CUDA/ROCm toolkit so should always be around if building MFEM/HYPRE with CUDA/HIP.

Reply:
I see you removed the auto-detection of `HYPRE_USING_HIP` -- this is not what I was suggesting and I think we want to keep that. Why do you want to remove it?

Reading your last comment, I thought you basically added auto-detection for `HYPRE_USING_CUDA` but I guess I misunderstood your comment.

Reply:
Resolved in [1354c29](https://github.com/mfem/mfem/pull/3973/commits/1354c299058a9348a8a32561bad31090a451b04b). Sorry about that, I thought we could get away with using the MFEM variables (`MFEM_USE_CUDA`/`MFEM_USE_HIP`) to guess how HYPRE was built. This is what was happening before for CUDA builds at least. But I agree the HIP handling was better so I've reverted to what was there before.

Reply:
Thanks @sebastiangrimberg! This looks good.

If we wanted to be super strict, we can detect specifically what CUDA/HIP libraries HYPRE was build with -- there are macros that specify that in `HYPRE_config.h`, e.g. `HYPRE_USING_CURAND`, `HYPRE_USING_CUSOLVER`, etc; see https://github.com/hypre-space/hypre/blob/master/src/config/HYPRE_config.h.in. However, I don't think it is crucial -- for now we assume default CUDA/HIP build of HYPRE which will work for non-default builds too.